### PR TITLE
Improvements to FLIP

### LINKED ISSUE #62: Improvements to FLIP
1) Add "reverse" as a third option to the faceCycle property of widgets. The widget should then be flipped to the _previous_ face instead of the next.

Reason: This is useful for multi-faced widgets like dice, as some games need to flip it to previous face rather than always to the next.

2) Allow FLIP to temporarily override the widgets faceCycle property by adding a faceCycle parameter:
```
{
    func: “Flip”,
    holder: "<holderId>",
    faceCycle: "random"
}
```

Reason: In some games, the widget's faceCycle setting may not be valid for all situations in which a FLIP should be made.

### PR DESCRIPTION
**Pull request by larkob**
Allows FLIP operation to override the faceCycle property of the items in the holder.

Introduces new faceCycle option `backward` (reverse direction).

Introduces collection parameter to FLIP operation.

Closes #62

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3216/pr-test (or any other room on that server)

**Comment by 96LawDawg** [comment]
This needs a little bit of tweaking.  It works fine for the card and the stick (from my Alleys game).  But it does not work well on the checker or the dice.  When it hits the last face, the room just flashes.  It needs to loop back around to the previous (highest) face.

**Comment by larkob** [comment]
Thanks, I fixed the issue.

**Review by 96LawDawg — APPROVED** [review]
Approved through testing.  Did not review code.

File: client/js/widgets/basicwidget.js [hunk 1]
```diff
@@ -63,11 +63,16 @@ class BasicWidget extends Widget {
...
   }
 
-  flip(setFlip) {
+  flip(setFlip, faceCycle) {
     if(setFlip !== undefined && setFlip !== null)
       this.p('activeFace', setFlip);
-    else
-      this.p('activeFace', Math.floor(this.p('activeFace') + (this.p('faceCycle') == 'random' ? Math.random()*99999 : 1)) % this.p('faces').length);
+    else {
+      const fC = (faceCycle == 'previous' || faceCycle == 'random') ? faceCycle : this.p('faceCycle');
```
Comment (ArnoldSmith86) [inline client/js/widgets/basicwidget.js:70]: - [ ] For `faceCycle == 'ordered'` this ignores the provided `faceCycle` and falls back to `this.p('faceCycle')` which could be something else.

File: client/js/widgets/basicwidget.js [hunk 2]
```diff
@@ -63,11 +63,16 @@ class BasicWidget extends Widget {
...
+      if (fC == 'previous')
+        this.p('activeFace', (this.p('activeFace') == 0) ? this.p('faces').length-1 : this.p('activeFace') -1);
```
Comment (ArnoldSmith86) [inline client/js/widgets/basicwidget.js:72]: Just FYI: the parenthesis around `this.p('activeFace') == 0` are not necessary.

**Review by ArnoldSmith86 — CHANGES_REQUESTED** [review]
I didn't test it but the code looks fine except for that one issue.

**Review by RaphaelAlvez — APPROVED** [review]
tested and approve

**Comment by larkob** [comment]
@ArnoldSmith86 @96LawDawg I've made a few changes to the code because I wanted to add the collection parameter to FLIP while I'm at it, so please take another look if you have time.

**Comment by RaphaelAlvez** [comment]
@larkob Tested and approved again. I downloaded the test room and I'm merging it as I said I would

**Pull request merged**